src/quadiron-c.cpp

```diff
@@ -0,0 +1,450 @@
+            new quadiron::fec::RsFnt<uint32_t>(
+                systematic ? quadiron::fec::FecType::SYSTEMATIC
+                           : quadiron::fec::FecType::NON_SYSTEMATIC,
+                word_size,
+                n_data,
+                n_parities,
+                pkt_size));
+    }
+
+    return nullptr;
+}
+
+/**
+ * Delete FEC
+ *
+ * @param fecp the FEC instance pointer
+ */
+void quadiron_fnt32_delete(struct QuadironFnt32* fecp)
+{
+    delete reinterpret_cast<quadiron::fec::RsFnt<uint32_t>*>(fecp);
+}
+
+/**
+ * Return metadata size
+ *
+ * we assume that a special value of 65536 may occur uniformly.
+ * we count 4 bytes per special value.
+ * we see large and roundup by 16 items.
+ */
+int quadiron_fnt32_get_metadata_size(
+    struct QuadironFnt32* /* fecp */,
+    int block_size)
+{
+    return ((block_size / 65536) + 16) * 4;
+}
+
+/**
+ * Encode blocks
+ *
+ * @param data must be exactly n_data (use nullptr when missing) -
+ * buffers must allocate block_size + metadata_size
+ * @param parity must be exactly n_parities otherwise
+ * @param missing_idxs array of missing_idxs of len code_len indicating presence
+ * (value 1) or absence of fragments (value 0)
+ * @param block_size the block size in bytes
+ *
+ * @return true if decode succeeded, else false
+ */
+int quadiron_fnt32_encode(
+    struct QuadironFnt32* fecp,
+    uint8_t** data,
+    uint8_t** parity,
+    int* missing_idxs,
+    int block_size)
+{
+    quadiron::fec::RsFnt<uint32_t>* fec =
+        reinterpret_cast<quadiron::fec::RsFnt<uint32_t>*>(fecp);
+    std::vector<uint8_t*> data_vec(fec->n_data, nullptr);
+    std::vector<uint8_t*> parities_vec(fec->n_outputs, nullptr);
+    std::vector<quadiron::Properties> parities_props(fec->n_outputs);
+    std::vector<int> missing_idxs_vec(
+        missing_idxs, missing_idxs + fec->code_len);
+    int metadata_size = quadiron_fnt32_get_metadata_size(fecp, block_size);
+
+    if (fec->type == quadiron::fec::FecType::SYSTEMATIC) {
+        for (unsigned i = 0; i < fec->n_data; i++) {
+            data_vec[i] = data[i] + metadata_size;
+        }
+        for (unsigned i = 0; i < fec->n_parities; i++) {
+            parities_vec[i] = parity[i] + metadata_size;
+        }
+    } else {
+        for (unsigned i = 0; i < fec->n_data; i++) {
+            data_vec[i] = data[i] + metadata_size;
+            parities_vec[i] = data[i] + metadata_size;
+        }
+        for (unsigned i = 0; i < fec->n_parities; i++) {
+            parities_vec[fec->n_data + i] = parity[i] + metadata_size;
+        }
+    }
+
+    fec->encode_blocks_vertical(
+        data_vec, parities_vec, parities_props, missing_idxs_vec, block_size);
+
+    if (fec->type == quadiron::fec::FecType::SYSTEMATIC) {
+        for (unsigned i = 0; i < fec->n_parities; i++) {
+            uint32_t* metadata = reinterpret_cast<uint32_t*>(parity[i]);
+            int ret = parities_props[i].to_buffer(metadata, metadata_size / 4);
+            if (-1 == ret) {
```

Comment:
I think we mostly use the natural order of `(ret == -1)` in the code base.

I know that Yoda's conditions have been used to prevent mistake such as `if (ret = -1)` but compilers been warning about this mistake for ages now, there is now no reason to make the code less natural to prevent it.